### ISSUE TITLE
Highlighting codepoints within a bracket range messes up rendering

### ISSUE THREAD

**Issue by lenormf** [mawww/kakoune#638]
Hi,

Just found the weirdest bug, I'm not sure if boost is to blame or not. Open a new buffer with the following text in it:

```
“We ought to scrape this planet clean of every living thing on it,”
```

Try to highlight the quotes with the following highlighter:
```
addhl regex '“|”' 0:red
```

Everything works as expected, now try with this:
```
addhl regex '[“”]' 0:red
```

The original text is now rendered with raw codepoint codes (`+80U+9c` and `+80U+9d`). Note that highlighting is done as expected, only the characters are not rendered.

**EDIT**: another example that has nothing to do with highlighting, create a multiple selection whose selections only contain the unicode quote characters used above (`“` or `”`), and use `'` to rotate through the selections to have the rendering issue reproduced. Not all characters have this issue though.

**Comment by mawww** [comment]
I think you are hitting boost not supporting codepoints but only bytes here:

* `“|”` appears to boost as `<utf8 byte seq>|<utf8 byte seq>` so it find matches properly even without considering codepoints.
* `[“”]` appears to boost as [<byte><byte><byte><byte>] and it find either of those bytes as a single match. As we now have display atoms that split code points (one for +80U, another one for +9c) we end up with the Kakoune invalid codepoint highlighter displaying them as raw codepoint codes.

**Comment by lenormf** [comment]
Regarding highlighting messing up, I can just replace the bracket ranges with the pipes, however cycling through selections that only contain those codepoints with `'` is more annoying. Is there any way to get this fixed ?

**Comment by mawww** [comment]
Cannot reproduce that second issue, but not sure exactly what I should do (`%s“|”<ret>` on the sample text you gave plus `''''` did not do anything).

**Comment by lenormf** [comment]
In a buffer that contains:
```
“”
```

Use `gg<a-l>s.<ret>'`

**Comment by mawww** [comment]
Oh, thats the same problem with `s.<ret>` boost selects each matches for `.`, which are bytes and not codepoint, that said, it might be better if kakoune auto corrected the selection to only contain full codepoints.

**Comment by lenormf** [comment]
For documenting purposes, this is being fixed in #639.
